File: testdata/fixtures/php/CWE-798.php
<?php
/* CWE-798: Use of Hard-coded Credentials.
 *
 * A secret in the source is in every deploy of it, in the repository's whole
 * history, and in any backup of the document root - which on a
 * misconfigured server is served as plain text. It cannot be rotated without
 * a release.
 */

/* ruleid: php-hardcoded-credential */
define('DB_PASSWORD', 'hunter2trustno1');
/* ruleid: php-hardcoded-credential */
const API_TOKEN = 'ghp_16C7e42F292c6912E7710c838347Ae178B4a';

class Connection
{
    /* ruleid: php-hardcoded-credential */
    private $password = 'correct-horse-battery';
    /* ruleid: php-hardcoded-credential */
    public static $apiKey = 'sk_live_51H8fJ2KpQ9vXnZ3mLb7T';

    public function connect($host)
    {
        /* ruleid: php-hardcoded-credential */
        $secret = 'aG9yc2ViYXR0ZXJ5c3RhcGxl';
        /* ruleid: php-hardcoded-credential */
        $config = ['user' => 'app', 'password' => 's3kr1t-passphrase'];

        /* ok: php-hardcoded-credential */
        $fromEnv = getenv('DB_PASSWORD');
        /* ok: php-hardcoded-credential */
        $empty = '';
        /* ok: php-hardcoded-credential */
        $prompt = 'Please enter your password';
        /* ok: php-hardcoded-credential */
        $format = 'password=%s';
        /* ok: php-hardcoded-credential */
        $redacted_password = 'PASSWORD';
        /* ok: php-hardcoded-credential */
        $api_key = 'YOUR_API_KEY_HERE';

        return [$secret, $config, $fromEnv, $empty, $prompt, $format,
                $redacted_password, $api_key];
    }
}

File: testdata/fixtures/c/CWE-798.c
/* CWE-798: Use of Hard-coded Credentials.
 *
 * A secret in the source is a secret in every binary built from it, in the
 * repository's whole history, and in the core dump. It cannot be rotated
 * without a release, and `strings` finds it.
 */
#include <stdlib.h>
#include <string.h>
#include <mysql.h>

/* ruleid: c-hardcoded-credential */
#define DB_PASSWORD "hunter2trustno1"
/* ruleid: c-hardcoded-credential */
#define API_TOKEN "ghp_16C7e42F292c6912E7710c838347Ae178B4a"
/* ruleid: c-hardcoded-credential */
static const char *aws_secret_key = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY";
/* ruleid: c-hardcoded-credential */
static char admin_passwd[] = "correct-horse-battery";

#define PASSWORD_MAX 64
#define RETRY_LIMIT 3

static const char *key_format = "%s:%s";
static const char *prompt = "enter your password";

int connect_db(MYSQL *conn, const char *host, const char *user)
{
    /* ruleid: c-hardcoded-credential */
    const char *passphrase = "s3kr1t-passphrase";
    /* ruleid: c-hardcoded-credential */
    mysql_real_connect(conn, host, user, "hunter2trustno1", "app", 0, NULL, 0);

    /* ok: c-hardcoded-credential */
    const char *from_env = getenv("APP_PASSWORD");
    /* ok: c-hardcoded-credential */
    char password[PASSWORD_MAX];
    /* ok: c-hardcoded-credential */
    const char *empty_secret = "";
    /* ok: c-hardcoded-credential */
    mysql_real_connect(conn, host, user, from_env, "app", 0, NULL, 0);

    (void)passphrase; (void)key_format; (void)prompt;
    (void)aws_secret_key; (void)admin_passwd;
    return password[0] + (from_env != NULL) + (empty_secret != NULL);
}
